### PR TITLE
FEAT: Complex Data Type Support- money & smallmoney

### PR DESCRIPTION
**Pull request by gargsaumya**
### Work Item / Issue Reference  

> [AB#34936](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/34936)
[AB#34937](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/34937)

> GitHub Issue: #<ISSUE_NUMBER>

-------------------------------------------------------------------
### Summary   
  
This pull request enhances support for SQL Server MONEY and SMALLMONEY types in the Python driver, ensuring correct handling of boundary values, NULLs, and round-trip conversions. It also improves the encoding and decoding of decimal values and adds comprehensive tests to validate the new behavior.

**MONEY and SMALLMONEY Type Handling:**

* Updated `_map_sql_type` in `cursor.py` to detect decimal values within MONEY and SMALLMONEY ranges and bind them as strings, falling back to generic numeric binding otherwise.

**Decimal Value Encoding/Decoding:**

* Improved encoding of decimal values in `ddbc_bindings.cpp` by writing them as little-endian byte arrays, ensuring compatibility with SQL_NUMERIC_STRUCT.
* Enhanced decoding of SQL DECIMAL/NUMERIC types to Python `decimal.Decimal`, with robust NULL and error handling during result fetching.

**Testing Enhancements:**

* Added comprehensive tests in `test_004_cursor.py` to verify MONEY and SMALLMONEY handling, including:
  - Insert/fetch of typical, boundary, and NULL values
  - Round-trip conversion checks
  - Boundary value assertions
  - Validation that out-of-range and invalid values raise errors

### CODE REVIEW

File: mssql_python/cursor.py [hunk 1]
```diff
@@ -282,18 +282,39 @@ def _map_sql_type(self, param, parameters_list, i):
...
-            )  # Replace the parameter with the dictionary
-            return (
-                ddbc_sql_const.SQL_NUMERIC.value,
-                ddbc_sql_const.SQL_C_NUMERIC.value,
-                parameters_list[i].precision,
-                parameters_list[i].scale,
-                False,
-            )
+        # Detect MONEY / SMALLMONEY range
+            if -214748.3648 <= param <= 214748.3647:
```
Comment (sumitmsft) [inline mssql_python/cursor.py]: The binding for MONEY/SMALLMONEY returns scale=0. SQL Server defines both with a fixed scale of 4 (SMALLMONEY precision 10, scale 4; MONEY precision 19, scale 4). Please correct precision/scale (and preferably bind as numeric rather than VARCHAR) to avoid implicit conversions and possible plan inefficiencies.

Just to give some context:
- the MONEY and SMALLMONEY paths return tuples with scale = 0 and a precision implicitly derived from len(str(param)), but SQL Server defines:
SMALLMONEY: precision 10, scale 4
MONEY: precision 19, scale 4

- Incorrect metadata can force SQL Server to treat the parameter as a string → implicit VARCHAR → MONEY conversion per execution. This may cause SQL query plan inefficiencies

File: mssql_python/cursor.py [hunk 2]
```diff
@@ -282,18 +282,39 @@ def _map_sql_type(self, param, parameters_list, i):
...
+                # smallmoney
+                parameters_list[i] = str(param)
+                return (
+                    ddbc_sql_const.SQL_VARCHAR.value,
+                    ddbc_sql_const.SQL_C_CHAR.value,
+                    len(parameters_list[i]),
```
Comment (sumitmsft) [inline mssql_python/cursor.py:298]: Precision is currently derived via len(str(param)), which is incorrect. Precision = total number of digits (excluding sign and decimal point). Please use fixed MONEY/SMALLMONEY precisions (19 / 10) or compute digit count from the Decimal tuple instead of string length

File: mssql_python/cursor.py (hunk 1 above)
Comment (sumitmsft) [inline mssql_python/cursor.py]: Boundary checks for MONEY/SMALLMONEY currently use float literals (e.g. -214748.3648) against Decimal parameters. This introduces an unnecessary float→Decimal conversion with possible rounding artifacts at exact financial limits (especially the large MONEY bounds).

Can we pre-declare boundary constants using precise Decimal string constructors:

MONEY_MIN = Decimal('-922337203685477.5808')
MONEY_MAX = Decimal('922337203685477.5807')
SMALLMONEY_MIN = Decimal('-214748.3648')
SMALLMONEY_MAX = Decimal('214748.3647')

This way we compare only Decimal to Decimal.

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 3]
```diff
@@ -474,15 +474,16 @@ SQLRETURN BindParameters(SQLHANDLE hStmt, const py::list& params,
...
                 decimalPtr->scale = decimalParam.scale;
                 decimalPtr->sign = decimalParam.sign;
-                // Convert the integer decimalParam.val to char array
-                std::memset(static_cast<void*>(decimalPtr->val), 0, sizeof(decimalPtr->val));
-                std::memcpy(static_cast<void*>(decimalPtr->val),
-			    reinterpret_cast<char*>(&decimalParam.val),
-                            sizeof(decimalParam.val));
+                // Zero out the byte array
+                std::memset(decimalPtr->val, 0, sizeof(decimalPtr->val));
+
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp]: This code copies the unscaled integer `decimalParam.val` into the SQL_NUMERIC_STRUCT buffer without first confirming that:

- The digit count of the unscaled value fits the declared precision. `digits <= precision`
- `scale <= precision`
- There’s no early exit with a clear error if the value exceeds precision leading to potential silent mismatch between declared metadata and actual value.

While current uint64_t values won’t overflow the 16‑byte array, a value with more digits than precision will still be sent and this could be a data integrity risk for numeric correctness.

For this we can some tests as well:

- A test that constructs a NumericData whose val has more digits than precision and assert it raises / returns error.

**Review by sumitmsft — CHANGES_REQUESTED** [review]
Left a few comments

**Review by sumitmsft — CHANGES_REQUESTED** [review]
Left a few comments

File: mssql_python/cursor.py (hunk 2 above)
Comment (bewithgaurav) [reply mssql_python/cursor.py:298]: just correcting the context here - the len(parameters_list[i]) is column size, not precision 
`map_sql_type` returns the below
- `A tuple containing the SQL type, C type, column size, and decimal digits.`
@gargsaumya the column size will be length only right, please correct if I'm wrong
the tests look correct and they're passing for precision cases

**Review by bewithgaurav — APPROVED** [review]
